### PR TITLE
[wsi] Refactor platform system to support multiple WSI implementations

### LINKED ISSUE #3321: Native builds with SDL and GLFW are incompatible, so should have different SONAMEs
As far as I can tell, each of the available WSIs for DXVK Native has a different ABI: in the SDL2 WSI, whenever a function in the DirectX API returns or receives a `HWND` it's really a `SDL_Window *`, but in the GLFW WSI, whenever a function returns or receives a `HWND` it's really a `GLFWwindow *`.

When SDL 3 becomes stable, presumably there will be SDL3 games that want to use DXVK Native, at which point that would become a third incompatible WSI?

If that's the case, then they should have different SONAMEs. One option would be to append the WSI to the `dxvk_name_prefix`, making them `libdxvk_sdl2_d3d9.so`, `libdxvk_glfw_d3d9.so` and so on. `libdxvk_d3d9.so` could be a symlink to `libdxvk_sdl2_d3d9.so`, and so on, for backwards compatibility.

Or, the default `sdl2` WSI could use an unadorned name because it got there first, and all other WSIs (including SDL3, GLFW, and any other WSIs that are invented in future) could append their name to the `dxvk_name_prefix`?

### PR DESCRIPTION
**Pull request by flibitijibibo**
This reorganizes the WSI backend system to support multiple backends in a single binary - this is important for dxvk-native which can support either SDL2 or GLFW. We achieve this by doing the following:

- Add a refcounted init/quit system to WSI, which is used by DxvkInstance
- Move the WSI-specific code from dxvk_platform_exts to WSI
- SDL2 and GLFW are now dynamically _loaded_, rather than dynamically _linked_, to avoid hard dependencies that could otherwise be optional
- The WSI backends are now implementations of a common interface, which effectively implements the same API that was exposed to the rest of DXVK. WSI init will iterate through each compiled-in backend and continue until one of them successfully initializes. This can be overridden with the `DXVK_WSIDRIVER` environment variable.

Supercedes #3736

Part of #3451

Fixes #3321

Obsoletes #3324

### CODE REVIEW

**Comment by flibitijibibo** [comment]
Latest push is another rebase - I've got some spare cycles this week so if there's anything that needs to be addressed in this (or the other two) PRs let me know and I'll get it done ASAP!

File: src/wsi/wsi_platform.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+  void init() {
+    if (s_refcount++ > 0)
+      return;
+
+    bool success = false;
+    const std::string hint = dxvk::env::getEnvVar("DXVK_WSIDRIVER");
+    for (const WsiBootstrap *b : wsiBootstrap) {
+      if ((hint != "") && (hint != b->name))
+          continue;
```
Comment (doitsujin) [inline src/wsi/wsi_platform.cpp:31]: Talking to @Joshua-Ashton, I think a potential issue here is that if multiple WSIs are available we might initialize a random one that the app might not be able to deal with if the env var isn't set. Maybe we should outright fail if we have no env var, or just default to the first WSI in the list?

**Review by doitsujin — COMMENTED** [review]
Looking into (finally) getting this merged for the 2.4 release among other stuff, looks good overall, just one concern.

File: src/wsi/wsi_platform.cpp (hunk 1 above)
Comment (flibitijibibo) [reply src/wsi/wsi_platform.cpp:31]: I'd be okay with that; it's something I've run into particularly with SDL3 migration - will make this change later this morning!

**Comment by flibitijibibo** [comment]
Latest push adjusts the behavior of DXVK_WSIDRIVER, particularly when the variable is unset: When Win32 is available we assume that is the default, when it is _not_ available we throw an explicit error requesting that the variable be set.

**Review by doitsujin — APPROVED** [review]

**Pull request merged**